### PR TITLE
[Datasets] [Prototype] Expose new `.to_torch_datapipe` API that leverages and returns `torchdata.IterDataPipe`.

### LINKED ISSUE #23199: [Datasets] Port `.to_torch()` to new `IterDataPipe` API.
[`torchdata`](https://github.com/pytorch/data) has a new `IterDataPipe` API that will subsume the old `IterableDataset` API, which is now considered the legacy Torch data interface. We should port our `.to_torch()` API to this new interface for a few reasons:
1. Stay current with the canonical Torch data interface in order to provide a better experience to end-users.
2. Stay open to better interoperability with Torch's data loading utilities and abstractions.

### PR DESCRIPTION
**Pull request by clarkzinzow**
This PR ports our `.to_torch()` API to the new `IterDataPipe` API that's currently in beta (see [`torchdata`](https://github.com/pytorch/data)), since the old `IterableDataset` API is now considered the legacy Torch dataset interface.

## TODOs

- [x] Figure out Torch version story. `torchdata` currently [requires](https://github.com/pytorch/data#version-compatibility) `torch==1.11.0`, but that seems like a strict version requirement to impose on our users. We could (a) fall back to the `IterableDataset` API for `torch<1.11.0`, (b) expose a separate `to_torch_datapipe()` API that fails for `torch<1.11.0`, or (c) wait until `torchdata` has  a better backwards compatibility story. I'm leaning towards (b).

## Related issue number

Closes #23199

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by ericl** [comment]
@richardliaw can you find a ML reviewer for this?

File: python/requirements/ml/requirements_dl.txt [hunk 1]
```diff
@@ -10,11 +10,11 @@ keras==2.6.0
 tensorflow==2.6.0
 tensorflow-probability==0.14.0
 
-torch==1.9.0;sys_platform=="darwin"
-torchvision==0.10.0;sys_platform=="darwin"
+torch==1.11.0;sys_platform=="darwin"
```
Comment (amogkam) [inline python/requirements/ml/requirements_dl.txt:13]: If you make changes here can you also make the corresponding changes to `requirements_ml_docker.txt`? Thanks!

**Comment by msaroufim** [comment]
@clarkzinzow should we revisit or close this PR? Let me know if there's anything I can help with

**Comment by clarkzinzow** [comment]
Hey @msaroufim, thanks for the ping!

This has been put on the backburner for now since our user feedback for these kind of declarative/more complex `ds.to_torch()` and `ds.to_tf()` APIs was that they were too difficult to use:
- the batch structure semantics are complex, depending on multiple optional arguments (`feature_columns` and `label_column`) and their structure,
- feature dimension unsqueezing happening by default for single-feature datasets has surprised users,
- users have defaulted to trying to use a Torch `DataLoader` with the returned `IterableDataset`, which breaks e.g. batching and may result in redundant and/or inefficient in-iterator operations.

For now, we've opted for simpler `ds.iter_torch_batches()`/`ds.iter_tf_batches()` APIs that return a plain iterator of tensors (for single-column datasets) or dicts of tensors (for multi-column datasets).

We _do_ plan to circle back to more featureful integrations with ML framework dataset abstractions (such as `IterDataPipe`) in the next few months, after Ray 2.0. I think that it would be useful to sync at some point in the next few weeks to discuss what we think would make this integration a big value-add for us, such as being able to plug in Ray Datasets as the execution engine for a composed chain of `IterDataPipes`.

**Pull request closed without merging**